MINOR: fix NamedCache metrics in Streams

vvcephei opened the pull request:
* Fixes a bug in which all NamedCache instances in a process shared
one parent metric.

* Also fixes a bug which incorrectly computed the per-cache metric tag
(which was undetected due to the former bug).

* Drop the `StreamsMetricsConventions#xLevelSensorName` convention
in favor of `StreamsMetricsImpl#xLevelSensor` to allow `StreamsMetricsImpl`
to track thread- and cache-level metrics, so that they may be cleanly declared
from anywhere but still unloaded at the appropriate time. This was necessary
right now so that the `NamedCache` could register a thread-level parent sensor
to be unloaded when the thread, not the cache, is closed.

* The above changes made it mostly unnecessary for the `StreamsMetricsImpl` to
expose a reference to the underlying `Metrics` registry, so I did a little extra work
to remove that reference, including removing inconsistently-used and unnecessary
calls to `Metrics#close()` in the tests.

The existing tests should be sufficient to verify this change.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vvcephei committed:
MINOR: fix NamedCacheMetrics

* the parent sensor never got removed
* both the parent and child sensor names weren't scoped, so all hit
  ratio sensors in all caches were actually the same metric. This
  would cause a concurrency bottleneck in addition to reporting
  incorrect values.

vvcephei committed:
register parent for unloading at the thread level

vvcephei committed:
fix bug in determining store name from cache name

vvcephei committed:
clean up unnecessary metrics close

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsConventions.java:
```diff
@@ -1,39 +0,0 @@
-import java.util.LinkedHashMap;
-import java.util.Map;
-
-public final class StreamsMetricsConventions {
```
This logic is moved to `StreamsMetricsImpl`

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:
```diff
@@ -31,35 +31,90 @@
+                                          final Sensor.RecordingLevel recordingLevel,
+                                          final Sensor... parents) {
+        synchronized (threadLevelSensors) {
+            final String fullSensorName = threadName + "." + sensorName;
```
We previously prefixed this with "thread.". Was that necessary?

guozhangwang replied:
Probably not, and since sensor names are only used for internal bookkeeping there should be no compatibility issues with the change.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java:
```diff
@@ -355,45 +355,65 @@ private void update(final LRUCacheEntry entry) {
+            );
+            final Sensor threadLevelHitRatioSensor = metrics.threadLevelSensor("hitRatio", Sensor.RecordingLevel.DEBUG);
+            threadLevelHitRatioSensor.add(
+                new MetricName("hitRatio-avg", "stream-record-cache-metrics", "The average cache hit ratio.", allMetricTags),
```
The only difference between constructing the MetricName outright and using the registry to make it is that the registry's version will augment your tags with tags from the configs. I verified in https://github.com/apache/kafka/pull/4917/files#diff-06f3e8d1a2ac04162e76b98924c14b70R679 that we don't set tags in the configs, so it should be exactly the same to just construct it directly... right?

guozhangwang replied:
That is right, and originally we use `Metrics.metricName()` to leverage on the most common configs which is `"client-id" -> threadName`. But here you have removed it. Is that intentional? I think for thread-level we should have just one tag: `"client-id" -> threadName`, and for task-level we should have two tags: the one with thread level plus the task id, and for cache / store / processor-node we should have three tags, the two from task-level plus the record-cache-id / store-name / processor-node-name.

vvcephei replied:
That's right.

Unfortunately, though, we can't use Metrics to fill in any of those tags, because Metrics can only do global tags. As we have only one Metrics instance for the whole application, it can't even fill in the thread-level tags.

Since we maintain the state necessary to fill in all our own tags when we create metrics, there's no reason to use `Metrics#metricName` and it creates an unnecessary dependency of having a Metrics registry in scope. the `MetricName` constructor, on the other hand, creates no extra dependency. Given two equivalent methods of creating the metric name, it seems better to use the one with less baggage.

bbejeck replied:
True, will we ever want to have this ability?  But the change seems fine to me.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/ThreadCache.java:
```diff
@@ -101,7 +101,7 @@ public static String taskIDfromCacheName(final String cacheName) {
      */
     public static String underlyingStoreNamefromCacheName(final String cacheName) {
-        String[] tokens = cacheName.split("-");
+        String[] tokens = cacheName.split("-", 2);
```
`cacheName` is created in `nameSpaceFromTaskIdAndStore` in L84 by concatenating the `taskID` and the `underlyingStoreName` with a `"-"`. So when we take them back apart, we should just split on hyphen *once*. Previously, this would split on *every* hyphen, returning only the part of the store name up to the first hyphen. This is a problem since store names may contain hyphens.

The result is that all named caches for stores with the same prefix (like "agg-store" and "agg-STATE_STORE_000000") will both try to create metrics with a storeName of just "agg".

We never noticed before because all NamedCaches were actually just creating one sensor, and it's ok to register duplicate metrics on the same sensor. Once I split the sensors up, it threw an exception on the duplicate metrics.

guozhangwang replied:
Great catch, thanks @vvcephei !

bbejeck replied:
seems reasonable to me

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/ThreadCache.java:
```diff
@@ -91,7 +91,7 @@ public static String nameSpaceFromTaskIdAndStore(final String taskIDString, fina
      */
     public static String taskIDfromCacheName(final String cacheName) {
-        String[] tokens = cacheName.split("-");
+        String[] tokens = cacheName.split("-", 2);
```
This isn't actually a problem, but I did it for consistency with the other change.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java:
```diff
@@ -121,7 +121,6 @@ private void initStore(final boolean enableCaching) {
 
     @After
     public void closeStore() {
-        context.close();
```
These are unnecessary. See my comment on the target method below.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -243,17 +243,8 @@ private StreamThread createStreamThread(@SuppressWarnings("SameParameterValue")
-        assertNotNull(metrics.getSensor(defaultPrefix + ".punctuate-latency"));
-        assertNotNull(metrics.getSensor(defaultPrefix + ".task-created"));
-        assertNotNull(metrics.getSensor(defaultPrefix + ".task-closed"));
-        assertNotNull(metrics.getSensor(defaultPrefix + ".skipped-records"));
```
Do we need to verify that the sensors got created? It seems like what we care about is that the metrics get created.

Removing this allows us to alter the sensor names without breaking tests.

guozhangwang replied:
Sounds good.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/state/internals/NamedCacheTest.java:
```diff
@@ -83,18 +83,15 @@ public void testMetrics() {
         metricTags.put("task-id", taskIDString);
         metricTags.put("client-id", "test");
 
-        assertNotNull(streamMetrics.registry().getSensor("hitRatio"));
```
See my earlier comment on not testing for sensor names for the rationale.

bbejeck replied:
So my only question is if we can't or didn't create the metric we would not be able to create a sesnsor? I'm guessing the answer is yes, so the changes for sensor testing makes sense to me.

vvcephei replied:
Right, the workflow is:
1. create a sensor
2. add metrics to the sensor

So, if we didn't create the sensor, we couldn't add the metrics.

And if we don't add the metrics for this reason or any other, the tests will catch it, because they are verifying the metrics exist.

vvcephei commented on streams/src/test/java/org/apache/kafka/test/KStreamTestDriver.java:
```diff
@@ -147,7 +147,12 @@ protected void after() {
+            } catch (final RuntimeException e) {
+                new RuntimeException("Fatal exception initializing store.", e).printStackTrace();
+                throw e;
+            }
```
If store.init throws an exception, it fails the tests, but doesn't tell you why. I added this after a fair amount of debugging revealed that this line was throwing. WDYT about leaving it in?

guozhangwang replied:
When will an exception be thrown? Note that we are going to remove `KStreamTestDriver` soon.

vvcephei replied:
The particular exception I ran into was trying to register duplicate metrics, related to the cache name not being split correctly.

Either way; It was handy to have this here, so I wanted to give you the option of leaving it in place.

bbejeck replied:
Seems reasonable to me. But this raises a secondary thought will we be keeping `KStreamTestDriver` with the release of our new testing utils?

guozhangwang replied:
Sounds reasonable to me. Is there a similar issue with `TopologyTestDriver` as well?

vvcephei replied:
Possibly. I'll take a look.

vvcephei replied:
@bbejeck There is a ticket underway to use TopologyTestDriver in our internal tests (https://issues.apache.org/jira/browse/KAFKA-6474), but it's not clear at this point whether we will actually be able to remove KStreamTestDriver without polluting TopologyTestDriver.

vvcephei replied:
@guozhangwang , `TopologyTestDriver` uses the same store initialization code path that `KafkaStreams` does, so I'll leave it alone for now. If there's a visibility problem, we'll surely find out.

vvcephei commented on streams/src/test/java/org/apache/kafka/test/InternalMockProcessorContext.java:
```diff
@@ -306,10 +304,6 @@ public void restore(final String storeName, final Iterable<KeyValue<byte[], byte
     }
 
-    public void close() {
-        metrics.close();
```
`Metrics.close()` doesn't do anything we care about in tests, so this just adds unnecessary complexity and an unnecessary dependency on `StreamsMetricsImpl#registry()`. I elected just to remove it; WDYT?

guozhangwang replied:
Sounds good to me.

bbejeck replied:
seems reasonable, but since `metrics` is an instance field on `AbstractProcessorContext` do we at least want to provide access to the `metrics` fields for any future tests?

vvcephei replied:
I'm not sure I followed your question completely, but the metrics field in APC isn't a reference to the Metrics registry, just a Map<MetricName,Metric> list of all metrics.

Also, future tests are free to add any references they need. I'd prefer to eagerly drop unnecessary code to keep complexity low and let future me worry about whether he needs to introduce more complexity in support of some feature.

vvcephei reviewed (commented):
I made a pass to explain some rationale further.

vvcephei committed:
remove unused imports

vvcephei committed:
remove unused imports

vvcephei commented:
retest this, please.

vvcephei commented:
Just the jdk7 test failed with one of the producer tests.

vvcephei commented:
Oh, forgot to request reviews! @guozhangwang @mjsax @bbejeck , please let me know what you think!

Thanks,
-John

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImpl.java:
```diff
@@ -31,35 +31,90 @@
+                                          final Sensor.RecordingLevel recordingLevel,
+                                          final Sensor... parents) {
+        synchronized (threadLevelSensors) {
+            final String fullSensorName = threadName + "." + sensorName;
```
Why not organizing the thread-level sensors as cache-level sensors as well? I.e. `Map<String, Deque<String>> threadLevelSensors = new HashMap<>()` where the string key is just `threadName`, since we will only remove sensors for the whole thread at once.

vvcephei replied:
I started to do it that way, but I realized that right now StreamsMetricsImpl is only created and used in the thread level (its constructor takes a thread name), so that map would only ever have one key. It seemed needless to introduce that extra machinery for thread-level metrics if it's not necessary, and in fact it would be misleading, since it would suggest that the SMI was somehow managing sensors for more than one thread.

I figured that if, in the future, we wanted to make SMI a singleton for the whole app, we could easily retrofit threadLevelSensor.

It might be better as a thread-level class, though, so we don't have to worry about contention.

guozhangwang replied:
Makes sense, we can do that later.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java:
```diff
@@ -355,45 +355,65 @@ private void update(final LRUCacheEntry entry) {
+            );
+            final Sensor threadLevelHitRatioSensor = metrics.threadLevelSensor("hitRatio", Sensor.RecordingLevel.DEBUG);
+            threadLevelHitRatioSensor.add(
+                new MetricName("hitRatio-avg", "stream-record-cache-metrics", "The average cache hit ratio.", allMetricTags),
```
nit: use a single local string field for group name `"stream-record-cache-metrics"` and tag key `"record-cache-id"` and `"task-id"`?

vvcephei replied:
I can do that, but just to entertain the rationale...

There's no performance penalty for the duplication, so the only consideration is code clarity, and it's easier to read the code with fewer substituted variables.

Specifically, we can see the whole metric name on this one line, as opposed to having to look at this line, and then look up the values of any variables.

I think there's a stronger argument for de-duplication when the value is more complicated, as in the case of tags, which is why I used a variable for that one instead of duplicating them for every metric. Even so, with the other variables inlined, the mental overhead is reduced enough that it's not too bad having to reference that one variable.

guozhangwang replied:
My rationale is mainly around typos :) In the past I have some experience spending much time on troubleshooting a typo caused issue, since these issues are usually not well exposed in the exception messages, e.g. if you had a `stream-record-cach-metrics` in one of the lines it would be hard to find out..

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java:
```diff
@@ -355,45 +355,65 @@ private void update(final LRUCacheEntry entry) {
+                cacheName,
+                "hitRatio",
+                Sensor.RecordingLevel.DEBUG,
+                threadLevelHitRatioSensor
```
Is this designed to have thread-level be the parent of the cache-level? I think originally we want to have task-level be the parent of cache-level (but there is a bug for that so it may not actually be the case).

vvcephei replied:
Ah, I misread the prior code. I'll correct it to use the task-level as the parent (and no thread-level aggregation).

vvcephei replied:
Double "ah"... this should have been obvious from the fact that the tags contains the task name...

guozhangwang reviewed (commented):
Made a pass and left some comments.

vvcephei committed:
add taskLevelSensor and use it for NamedCache aggregation

vvcephei commented:
@guozhangwang,

In https://github.com/apache/kafka/pull/4917/commits/57bd80ef93cba1627dee5dbec1296028b79427f1, I have added `StreamsMetricsImpl#taskLevelSensor`, following all the same patterns.

I made the NamedCache parent sensor a `taskLevelSensor`

I also updated the `TaskMetrics` to use the new `taskLevelSensor`, following the pattern of flattening the metric construction.

I couldn't tell if it's possible for two threads to have tasks with the same TaskId, but if it is, this change fixes a similar bug in which all metrics for the same TaskId would be conflated.

In any case, it certainly fixes a bug in which the parent sensor (with taskId = "all") would be conflated across all threads, introducing contention (and producing the wrong value).

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -72,16 +78,56 @@
+            taskName = id.toString();
             this.metrics = metrics;
-            taskCommitTimeSensor = metrics.addLatencyAndThroughputSensor("task", name, "commit", Sensor.RecordingLevel.DEBUG);
+
```
By flattening this sensor definition, we increase the verbosity of the code, but also make it clearer to read.

Previously, it would take a fair amount of work to determine what metrics with what names would be created, as you have to take note of the parameters and then descend through at least two levels of indirection. Now, you can see all the metrics clearly laid out here.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -72,16 +78,56 @@
+
+            // first add the global operation metrics if not yet, with the global tags only
+            final Map<String, String> allTagMap = metrics.tagMap("task-id", "all");
+            final Sensor parent = metrics.threadLevelSensor("commit", Sensor.RecordingLevel.DEBUG);
```
This fixes a bug that conflates this sensor across all threads.

Now, we get one sensor per thread, and it is properly unloaded when the thread, not this particular task, is removed.

bbejeck reviewed (commented):
just a couple of minor comments overall looks good.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -72,16 +78,56 @@
+            final Map<String, String> allTagMap = metrics.tagMap("task-id", "all");
+            final Sensor parent = metrics.threadLevelSensor("commit", Sensor.RecordingLevel.DEBUG);
+            parent.add(
+                new MetricName("commit-latency-avg", "stream-task-metrics", "The average latency of commit operation.", allTagMap),
```
Personally I'm still a bit leaning towards making the groupName as a field inside `TaskMetrics` same for `ThreadMetrics`, since it is quite hard to debug a typo issue in this case, and more string constants, more likely we will hit this.

vvcephei replied:
Ok, I'll make the group name a constant in those scopes.

guozhangwang reviewed (commented):
Just one more comment about the groupName, otherwise LGTM.

vvcephei committed:
move metric group names to constants

vvcephei commented:
@guozhangwang (and @bbejeck @mjsax ),

I've moved the group names to constants. I think that's the last outstanding review concern. Have I missed anything you all wanted me to answer/address?

bbejeck reviewed (approved):
Thanks @vvcephei , LGTM
